src/commands/layout/push.ts

```diff
@@ -0,0 +1,110 @@
+import { Args, Flags } from "@oclif/core";
+
+import BaseCommand from "@/lib/base-command";
+import { KnockEnv } from "@/lib/helpers/const";
+import { formatError, formatErrors, SourceError } from "@/lib/helpers/error";
+import * as CustomFlags from "@/lib/helpers/flag";
+import { merge } from "@/lib/helpers/object";
+import { formatErrorRespMessage, isSuccessResp } from "@/lib/helpers/request";
+import { indentString } from "@/lib/helpers/string";
+import { spinner } from "@/lib/helpers/ux";
+import * as EmailLayout from "@/lib/marshal/email-layout";
+import { WithAnnotation } from "@/lib/marshal/shared/types";
+
+export default class EmailLayoutPush extends BaseCommand<
+  typeof EmailLayoutPush
+> {
+  static aliases = ["email-layout:push", "email_layout:push"];
+
+  static summary =
+    "Push one or more email layouts from a local file system to Knock.";
+
+  static flags = {
+    environment: Flags.string({
+      summary:
+        "Pushing an email layout is only allowed in the development environment",
+      default: KnockEnv.Development,
+      options: [KnockEnv.Development],
+    }),
+    all: Flags.boolean({
+      summary: "Whether to push all layouts from the target directory.",
+    }),
+    "layouts-dir": CustomFlags.dirPath({
+      summary: "The target directory path to find all layouts to push.",
+      dependsOn: ["all"],
+    }),
+    commit: Flags.boolean({
+      summary: "Push and commit the layout(s) at the same time",
+    }),
+    "commit-message": Flags.string({
+      summary: "Use the given value as the commit message",
+      char: "m",
+      dependsOn: ["commit"],
+    }),
+  };
+
+  static args = {
+    emailLayoutKey: Args.string({
+      required: false,
+    }),
+  };
+
+  async run(): Promise<void> {
+    const { flags } = this.props;
+
+    // 1. First read all layout directories found for the given command.
+    const target = await EmailLayout.ensureValidCommandTarget(
+      this.props,
+      this.runContext,
+    );
+
+    const [layouts, readErrors] = await EmailLayout.readAllForCommandTarget(
+      target,
+      { withExtractedFiles: true },
+    );
+
+    if (readErrors.length > 0) {
+      this.error(formatErrors(readErrors, { prependBy: "\n\n" }));
+    }
+
+    if (layouts.length === 0) {
+      this.error(`No layout directories found in ${target.context.abspath}`);
+    }
+
+    // 2.TODO: Validate the layouts before upserting
+
+    // 3. Finally push up each layout, abort on the first error.
+    spinner.start(`‣ Pushing`);
+
+    for (const layout of layouts) {
+      const props = merge(this.props, { flags: { annotate: true } });
+
+      // eslint-disable-next-line no-await-in-loop
+      const resp = await this.apiV1.upsertEmailLayout<WithAnnotation>(props, {
+        ...layout.content,
+        key: layout.key,
+      });
+
+      if (isSuccessResp(resp)) continue;
```

Comment:
For `workflows` we update the directory with the success payload after pushing, but I'm not sure if we need this for `layouts`. It looks a bit redundant to me, but let me know what do you think @thomaswhyyou .

Reply:
We should replicate the same behavior as we do for workflows, and write the updated payload back to the file.

It's because we have certain read only fields that are system driven and get updated when we successfully push. Right now it's more relevant for workflows because fields like `active` or `valid` will be updated, in addition to timestamps like `updated_at`. 

But in the future, we will probably expose a `sha` fingerprint on each versioned/upserted resource, and the CLI will be able to use that to figure out whether the data actually changed or not.

Reply:
OK I now understand why we need this, thanks for the explanation! https://github.com/knocklabs/knock-cli/pull/236/commits/c21f4f937064a3e2e5c767b19a7bc2249abe813c